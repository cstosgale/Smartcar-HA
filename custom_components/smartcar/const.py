from enum import StrEnum, auto

DOMAIN = "smartcar"
DEFAULT_NAME = "Smartcar"
API_HOST = "https://api.smartcar.com"

PLATFORMS = [
    "sensor",
    "switch",
    "lock",
    "device_tracker",
    "binary_sensor",
    "number",
]

OAUTH2_AUTHORIZE = "https://connect.smartcar.com/oauth/authorize"
OAUTH2_TOKEN = "https://auth.smartcar.com/oauth/token"
SMARTCAR_MODE = "live"


class Scope(StrEnum):
    READ_VEHICLE_INFO = auto()
    READ_VIN = auto()
    READ_BATTERY = auto()
    READ_CHARGE = auto()
    READ_ENGINE_OIL = auto()
    READ_FUEL = auto()
    READ_LOCATION = auto()
    READ_ODOMETER = auto()
    READ_SECURITY = auto()
    READ_TIRES = auto()
    CONTROL_CHARGE = auto()
    CONTROL_SECURITY = auto()


REQUIRED_SCOPES = [
    Scope.READ_VEHICLE_INFO,
    Scope.READ_VIN,
]

CONFIGURABLE_SCOPES = [scope for scope in Scope if scope not in REQUIRED_SCOPES]

DEFAULT_SCOPES = [
    Scope.READ_BATTERY,
    Scope.READ_CHARGE,
    Scope.READ_LOCATION,
    Scope.READ_ODOMETER,
    Scope.READ_SECURITY,
    Scope.READ_VEHICLE_INFO,
    Scope.READ_VIN,
    Scope.CONTROL_CHARGE,
]
