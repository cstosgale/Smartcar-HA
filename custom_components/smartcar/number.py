# custom_components/smartcar/number.py

import logging; from typing import Any # Shorten imports
from aiohttp import ClientResponseError
from homeassistant.exceptions import ConfigEntryAuthFailed, HomeAssistantError
from homeassistant.components.number import (NumberEntity, NumberEntityDescription, NumberMode)
from homeassistant.config_entries import ConfigEntry; from homeassistant.const import PERCENTAGE
from homeassistant.core import HomeAssistant
from homeassistant.helpers.entity_platform import AddEntitiesCallback
from homeassistant.helpers.update_coordinator import CoordinatorEntity
from homeassistant.helpers.config_entry_oauth2_flow import OAuth2Session
from .const import DOMAIN, API_BASE_URL_V2
from .coordinator import SmartcarVehicleCoordinator

_LOGGER = logging.getLogger(__name__)
ENTITY_DESCRIPTIONS: tuple[NumberEntityDescription,...] = (NumberEntityDescription(key="charge_limit", name="Charge Limit", icon="mdi:battery-charging-80", mode=NumberMode.BOX, native_min_value=50.0, native_max_value=100.0, native_step=1.0, native_unit_of_measurement=PERCENTAGE),)

async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry, async_add_entities: AddEntitiesCallback) -> None:
    entry_data = hass.data[DOMAIN][entry.entry_id]
    coordinators: dict[str, SmartcarVehicleCoordinator] = entry_data.get("coordinators", {})
    session = entry_data["session"]
    entities = []
    token_scopes = entry.data.get("token", {}).get("scope", "").split()
    if "control_charge" not in token_scopes: _LOGGER.warning("Missing 'control_charge' scope."); return
    for vin, coordinator in coordinators.items():
        if coordinator.last_update_success and coordinator.data:
            for description in ENTITY_DESCRIPTIONS:
                if description.key == "charge_limit":
                    limit_data = coordinator.data.get("charge_limit")
                    if limit_data is not None and "limit" in limit_data:
                        entities.append(SmartcarChargeLimitNumber(coordinator, session, entry, description))
    _LOGGER.info("Adding %d Smartcar number entities", len(entities))
    async_add_entities(entities)

class SmartcarChargeLimitNumber(CoordinatorEntity[SmartcarVehicleCoordinator], NumberEntity):
    # ... (__init__, native_value, async_set_native_value, available as before) ...
    _attr_has_entity_name = True
    def __init__(self, coord, session, entry, desc): super().__init__(coord); self.vin=coord.vin; self.vehicle_id=coord.vehicle_id; self.session=session; self.entry=entry; self.entity_description=desc; self._attr_unique_id=f"{self.vin}_{desc.key}"; self._attr_device_info={"identifiers":{(DOMAIN,self.vin)}}
    @property
    def native_value(self): data=self.coordinator.data; limit_data=data.get("charge_limit") if data else None; limit_frac=limit_data.get("limit") if limit_data else None; return round(limit_frac*100.0) if limit_frac is not None else None
    async def async_set_native_value(self, value): # Error handling kept
        limit_frac=max(0.5,min(1.0,value/100.0)); _LOGGER.info("Attempting to set charge limit to %.2f%% for %s",value,self.vin); url=f"{API_BASE_URL_V2}/vehicles/{self.vehicle_id}/charge/limit"; payload={"limit":limit_frac}
        try: resp=await self.session.async_request("post",url,json=payload); resp.raise_for_status(); _LOGGER.info("Set charge limit cmd OK for %s",self.vin) # Removed early refresh
        except ClientResponseError as e:
            if e.status in(401,403): _LOGGER.warning("Auth err [%s] setting limit for %s",e.status,self.vin); self.entry.async_start_reauth(self.hass)
            elif e.status==409: _LOGGER.warning("Conflict (409) setting limit for %s.",self.vin); raise HomeAssistantError("Conflict setting limit (plugged in?)") from e
            else: _LOGGER.error("HTTP err setting limit for %s: %s",self.vin,e); raise HomeAssistantError(f"API err {e.status}") from e
        except ConfigEntryAuthFailed as e: _LOGGER.warning("AuthFail setting limit for %s",self.vin); self.entry.async_start_reauth(self.hass)
        except Exception as e: _LOGGER.exception("Unexpected err setting limit for %s",self.vin); raise HomeAssistantError("Unexpected limit err") from e
    @property
    def available(self): return super().available and self.coordinator.data is not None and self.coordinator.data.get("charge_limit") is not None and "limit" in self.coordinator.data["charge_limit"]