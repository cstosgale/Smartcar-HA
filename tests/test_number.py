"""Test number entities."""

from unittest.mock import AsyncMock

from aiohttp import ClientResponseError
from homeassistant.components.number import (
    ATTR_VALUE,
    DOMAIN as NUMBER_DOMAIN,
    SERVICE_SET_VALUE,
)
from homeassistant.const import ATTR_ENTITY_ID
from homeassistant.core import HomeAssistant
from homeassistant.exceptions import ServiceValidationError
from homeassistant.helpers import device_registry as dr, entity_registry as er
import pytest
from pytest_homeassistant_custom_component.common import MockConfigEntry
from pytest_homeassistant_custom_component.test_util.aiohttp import AiohttpClientMocker
from syrupy.assertion import SnapshotAssertion

from . import MOCK_API_ENDPOINT, setup_integration

NO_ERROR = None.__class__


@pytest.mark.usefixtures("enable_all_entities")
@pytest.mark.parametrize(
    (
        "api_status",
        "api_status_slug",
        "target_state",
        "expected_state",
        "expected_raises",
        "api_calls",
    ),
    [
        (200, "success", 90, 90, NO_ERROR, 1),
        (409, "unreachable", 90, 80, NO_ERROR, 1),
        (401, "unauthroized", 90, 80, NO_ERROR, 1),
        (500, "server", 90, 80, ClientResponseError, 1),
        (None, None, 40, 80, ServiceValidationError, 0),
        (None, None, 110, 80, ServiceValidationError, 0),
    ],
)
@pytest.mark.parametrize("vehicle_fixture", ["unknown_make"])
async def test_charging_limit(
    hass: HomeAssistant,
    mock_config_entry: MockConfigEntry,
    aioclient_mock: AiohttpClientMocker,
    snapshot: SnapshotAssertion,
    vehicle: AsyncMock,
    api_status: int,
    api_status_slug: str,
    target_state: int,
    expected_state: int,
    expected_raises: Exception,
    api_calls: int,
    device_registry: dr.DeviceRegistry,
    entity_registry: er.EntityRegistry,
) -> None:
    """Test updating charging limit."""

    await setup_integration(hass, mock_config_entry)
    assert len(aioclient_mock.mock_calls) == 1

    aioclient_mock.post(
        f"{MOCK_API_ENDPOINT}/v2.0/vehicles/{vehicle['id']}/charge/limit",
        status=api_status,
        json={
            "message": "Some message related to the action unused by our code",
            "status": api_status_slug,
        },
    )

    try:
        await hass.services.async_call(
            NUMBER_DOMAIN,
            SERVICE_SET_VALUE,
            service_data={ATTR_VALUE: target_state},
            target={ATTR_ENTITY_ID: "number.smartcar_784n_charge_limit"},
            blocking=True,
        )
    except Exception as error:  # noqa: BLE001
        raised_error = error
    else:
        raised_error = None  # type: ignore[assignment]

    number_state = hass.states.get("number.smartcar_784n_charge_limit")
    assert number_state.state == str(expected_state)
    assert isinstance(
        raised_error,
        expected_raises,  # type: ignore[arg-type]
    )

    assert len(aioclient_mock.mock_calls) == 1 + api_calls
    assert [tuple(mock_call) for mock_call in aioclient_mock.mock_calls[1:]] == snapshot
