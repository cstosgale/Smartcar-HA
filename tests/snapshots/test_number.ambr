# serializer version: 1
# name: test_charging_limit[unknown_make-200-success-90-90-NoneType-1]
  list([
    tuple(
      'post',
      URL('http://test.local/v2.0/vehicles/a1d50709-3502-4faa-ba43-a5c7565e6a09/charge/limit'),
      dict({
        'limit': 0.9,
      }),
      dict({
        'authorization': 'Bearer mock-token',
      }),
    ),
  ])
# ---
# name: test_charging_limit[unknown_make-401-unauthroized-90-80-NoneType-1]
  list([
    tuple(
      'post',
      URL('http://test.local/v2.0/vehicles/a1d50709-3502-4faa-ba43-a5c7565e6a09/charge/limit'),
      dict({
        'limit': 0.9,
      }),
      dict({
        'authorization': 'Bearer mock-token',
      }),
    ),
  ])
# ---
# name: test_charging_limit[unknown_make-409-unreachable-90-80-NoneType-1]
  list([
    tuple(
      'post',
      URL('http://test.local/v2.0/vehicles/a1d50709-3502-4faa-ba43-a5c7565e6a09/charge/limit'),
      dict({
        'limit': 0.9,
      }),
      dict({
        'authorization': 'Bearer mock-token',
      }),
    ),
  ])
# ---
# name: test_charging_limit[unknown_make-500-server-90-80-ClientResponseError-1]
  list([
    tuple(
      'post',
      URL('http://test.local/v2.0/vehicles/a1d50709-3502-4faa-ba43-a5c7565e6a09/charge/limit'),
      dict({
        'limit': 0.9,
      }),
      dict({
        'authorization': 'Bearer mock-token',
      }),
    ),
  ])
# ---
# name: test_charging_limit[unknown_make-None-None-110-80-ServiceValidationError-0]
  list([
  ])
# ---
# name: test_charging_limit[unknown_make-None-None-40-80-ServiceValidationError-0]
  list([
  ])
# ---
